File: leetcode/recursion/climbStairs/climbStairs.py
def climbStairs(n: int) -> int:
    cache = {}
    
    def cached(n):
        if n in cache:
            return cache[n]
        if n == 0:
            ans = 1
        elif n < 0:
            ans = 0
        else:
            # if one step
            ans = cached(n-1) + cached(n-2)
        cache[n] = ans
        return ans
            
            # if 2 step
    return cached(n)

File: leetcode/recursion/climbStairs/climbStairs.java
class Solution {
    HashMap<Integer, Integer> cache = new HashMap<Integer, Integer>();

    public int climbStairs(int n) {
        if (cache.containsKey(n)) {
            return cache.get(n);
        }
        
        int ans;
        if (n < 0) {
            ans = 0;
        } else if (n == 0) {
            ans = 1;
        } else {
            ans = climbStairs(n-1) + climbStairs(n-2);
        }
        
        cache.put(n, ans);
        return ans;
        
    }
}